src/librustdoc/passes/collect_intra_doc_links/early.rs

```diff
@@ -1,75 +1,148 @@
                     x.path_str
                 } else {
                     continue;
                 };
-                self.resolver.borrow_mut().access(|resolver| {
-                    let _ = resolver.resolve_str_path_error(
-                        span,
-                        &path_str,
-                        TypeNS,
-                        parent_module.unwrap_or(self.current_mod.to_def_id()),
-                    );
-                });
+                let _ = self.resolver.resolve_str_path_error(
+                    span,
+                    &path_str,
+                    Namespace::TypeNS,
+                    module_id,
+                );
+            }
+        }
+    }
+
+    /// Exports are replaced with item which they refer to, those items may have links in their
+    /// doc comments, those links are resolved at the item definition site, so we need to know
+    /// traits in scope at that definition site.
+    fn process_exports(&mut self, module_id: DefId) {
+        if !self.visited_mods.insert(module_id) {
+            return; // avoid infinite recursion
+        }
+
+        for export in self.resolver.module_exports(module_id) {
+            if export.vis == Visibility::Public {
+                if let Some(def_id) = export.res.opt_def_id() {
+                    if let Some(module_id) = self.resolver.parent(def_id) {
+                        self.add_traits_in_scope(module_id);
+                    }
+                }
+                if let Res::Def(DefKind::Mod, module_id) = export.res {
+                    self.process_exports(module_id);
+                }
             }
         }
     }
 }
 
-impl ast::visit::Visitor<'_> for IntraLinkCrateLoader {
+impl Visitor<'_> for IntraLinkCrateLoader<'_, '_> {
     fn visit_item(&mut self, item: &ast::Item) {
-        use rustc_ast_lowering::ResolverAstLowering;
-
-        if let ast::ItemKind::Mod(..) = item.kind {
-            let new_mod =
-                self.resolver.borrow_mut().access(|resolver| resolver.local_def_id(item.id));
-            let old_mod = mem::replace(&mut self.current_mod, new_mod);
+        if let ItemKind::Mod(..) = item.kind {
+            let old_mod = mem::replace(&mut self.current_mod, self.resolver.local_def_id(item.id));
 
             self.load_links_in_attrs(&item.attrs, item.span);
-            ast::visit::walk_item(self, item);
+            self.process_exports(self.current_mod.to_def_id());
+            visit::walk_item(self, item);
 
             self.current_mod = old_mod;
         } else {
             self.load_links_in_attrs(&item.attrs, item.span);
-            ast::visit::walk_item(self, item);
+            visit::walk_item(self, item);
         }
     }
+
+    fn visit_assoc_item(&mut self, item: &ast::AssocItem, ctxt: AssocCtxt) {
+        self.load_links_in_attrs(&item.attrs, item.span);
+        visit::walk_assoc_item(self, item, ctxt)
+    }
```

Comment:
Ooh good catch, this would be good to land even if we weren't precalculating traits.